e2e/ui/tests/download-logs.spec.js

```diff
@@ -0,0 +1,58 @@
+import { test, expect } from "@playwright/test";
+import { client, NOMAD_ADDR } from "../api-client.js";
+import { jobSpec } from "../input/jobs/service-discovery/service-discovery.js";
+import { token } from "../input/tokens/operator.js";
```

Comment:
Oh boy! Let me figure out how to rectify this issue.

Reply:
@tgross this issue is the REST API throwing an error. When we run this command, we're hitting the REST API. What I'm noticing is that we're hitting the wrong `url` and we might not properly be setting `NOMAD_ADDR`.

Reply:
The error is occurring because the HTTP request is refused:  `AxiosError: connect ECONNREFUSED 127.0.0.1:4200
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1187:16)`;

I'm trying to debug the error, but pretty blocked because I can't grok why the request gets a connection refused error when running `run.sh` and it doesn't happen when I run `npx playwright test`.

Below is the `reason` sent back to us.

```
{
  port: 4200,
  address: '127.0.0.1',
  syscall: 'connect',
  code: 'ECONNREFUSED',
  errno: -111,
  config: {
    transitional: {
      silentJSONParsing: true,
      forcedJSONParsing: true,
      clarifyTimeoutError: false
    },
    adapter: [Function: httpAdapter],
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 0,
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    env: { FormData: [Function] },
    validateStatus: [Function: validateStatus],
    headers: {
      Accept: 'application/json, text/plain, */*',
      'Content-Type': 'application/json',
      'X-Nomad-Token': '00e35c7c-4360-594f-317d-3738080f6889',
      'User-Agent': 'axios/0.27.2',
      'Content-Length': 111
    },
    url: 'http://localhost:4200/v1/namespace/prod',
    method: 'post',
    data: '{"Name":"prod","Description":"Production API Servers","Meta":{"contact":"platform-eng@example.com"},"Quota":""}'
  },
  request: <ref *1> Writable {
    _writableState: WritableState {
      objectMode: false,
      highWaterMark: 16384,
      finalCalled: false,
      needDrain: false,
      ending: false,
      ended: false,
      finished: false,
      destroyed: false,
      decodeStrings: true,
      defaultEncoding: 'utf8',
      length: 0,
      writing: false,
      corked: 0,
      sync: true,
      bufferProcessing: false,
      onwrite: [Function: bound onwrite],
      writecb: null,
      writelen: 0,
      afterWriteTickInfo: null,
      buffered: [],
      bufferedIndex: 0,
      allBuffers: true,
      allNoop: true,
      pendingcb: 0,
      constructed: true,
      prefinished: false,
      errorEmitted: false,
      emitClose: true,
      autoDestroy: true,
      errored: null,
      closed: false,
      closeEmitted: false,
      [Symbol(kOnFinished)]: []
    },
    _events: [Object: null prototype] {
      response: [Function: handleResponse],
      error: [Function: handleRequestError],
      socket: [Function: handleRequestSocket]
    },
    _eventsCount: 3,
    _maxListeners: undefined,
    _options: {
      maxRedirects: 21,
      maxBodyLength: 10485760,
      protocol: 'http:',
      path: '/v1/namespace/prod',
      method: 'POST',
      headers: [Object],
      agent: undefined,
      agents: [Object],
      auth: undefined,
      hostname: 'localhost',
      port: '4200',
      nativeProtocols: [Object],
      pathname: '/v1/namespace/prod'
    },
    _ended: false,
    _ending: true,
    _redirectCount: 0,
    _redirects: [],
    _requestBodyLength: 111,
    _requestBodyBuffers: [ [Object] ],
    _onNativeResponse: [Function (anonymous)],
    _currentRequest: ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 7,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: false,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      maxRequestsOnConnectionReached: false,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: true,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: null,
      _hasBody: true,
      _trailer: '',
      finished: false,
      _headerSent: true,
      _closed: false,
      socket: [Socket],
      _header: 'POST /v1/namespace/prod HTTP/1.1\r\n' +
        'Accept: application/json, text/plain, */*\r\n' +
        'Content-Type: application/json\r\n' +
        'X-Nomad-Token: 00e35c7c-4360-594f-317d-3738080f6889\r\n' +
        'User-Agent: axios/0.27.2\r\n' +
        'Content-Length: 111\r\n' +
        'Host: localhost:4200\r\n' +
        'Connection: close\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: nop],
      agent: [Agent],
      socketPath: undefined,
      method: 'POST',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      path: '/v1/namespace/prod',
      _ended: false,
      res: null,
      aborted: false,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: false,
      host: 'localhost',
      protocol: 'http:',
      _redirectable: [Circular *1],
      [Symbol(kCapture)]: false,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype]
    },
    _currentUrl: 'http://localhost:4200/v1/namespace/prod',
    [Symbol(kCapture)]: false
  }
}
```

Reply:
> I'm trying to debug the error, but pretty blocked because I can't grok why the request gets a connection refused error when running run.sh and it doesn't happen when I run npx playwright test.

Oh! I suspect that's because you're running tests against your dev server, which runs on port 4200, and not a build of Nomad that includes the UI, which runs on port 4646. When we run E2E tests we run it against a real build of Nomad that has the embedded UI, not the dev server proxied to it. It _should_ be possible to configure tests to run against both scenarios, but it wasn't on my radar to do.

Reply:
Gotcha. What do you think should be my next steps in getting closer to resolving this error?

Reply:
* Remove any hard coded instances of `http://localhost:4200` and replace it with the `NOMAD_ADDR` from the environment.
* Build a Nomad with the embedded UI so you can verify your changes work.
* (Stretch goal) Maybe add a second environment variable that, if set, allows for splitting web UI requests to the Ember dev server and not the `NOMAD_ADDR` that'll be serving the API.

Reply:
@tgross coming back to the original problem here of running the following code block and getting an connection request error from the hand rolled API client in the test suite:

```
$ export NOMAD_TOKEN=<redacted root token>
$ export NOMAD_ADDR=http://127.0.0.1:4646
$ nomad status
No running jobs

$ ./run.sh test
```

I believe this issue was the result of running the test suite in an isolated container that uses the [hand-rolled API client](https://github.com/hashicorp/nomad/pull/14216/commits/6ab80c39e779f83ceaf2775f689338a9da149aa4) to make a request to the Nomad server.

The challenge was that the reference to `localhost` in the isolated Docker container and your local machine were not the same. You did a great job setting the host in the network setting in the script, however, we need to specify the address for the Nomad Agent to bind to for network services (the HTTP interface -- now that we're using the hand-rolled API client*** to arrange and tear-down our test suite because we can't create and submit a job or namespace among other key operations within the Nomad UI).

So thankfully, I don't think that there's an issue with this test in particular, but your comment raises the need for thinking about the E2E UI more thoroughly on our end.

I think this PR is ready to merge into the trunk branch because this block of work is fine. However, this chunk of work isn't ready to make it into our actual test suite. I'd advocate for writing an RFC and have you as the approver so that we get ahead of issues like this in the future and we can make a bigger list of all the systems that we need to think about for these tests to safely (and not flakely) run:

- How will this work with Github Actions?
- Should we parallelize these tests because we're going to run into time-bound issues?
- Which considerations should we make in CircleCI?
- Do we need to propose changes to the Nomad Agent configuration for the E2E suite like binding the address?
- Tying into the previous question, what level of test coverage are we looking to accomplish? Do we want confidence in UI-specific features? Or do we want to test Nomad as a whole and fully mimic how our end users are configured and running Nomad?

Now that we've resolved that we can run the script successfully, is this PR ready to merge into the trunk branch?

_***: even if we could handle these actions in the Nomad UI. Having an API client still makes sense, IMO because the Nomad UI needs to test its real-time capability._

Reply:
> * How will this work with Github Actions?
> * Should we parallelize these tests because we're going to run into time-bound issues?
> * Which considerations should we make in CircleCI?
> * Do we need to propose changes to the Nomad Agent configuration for the E2E suite like binding the address?

When we run the E2E tests on nightly, the test runner runs on CircleCI and the cluster under test is running on AWS (see [README.md](https://github.com/hashicorp/nomad/blob/main/e2e/terraform/README.md)). So all the little details around the address binding are already taken care of and work just fine on CircleCI. We don't run E2E on GHA, but that's a conversation for later.

Because you're running on a real cluster, you _can_ parallelize the tests but if you do so you need to make sure the tests are safe to parallelize. i.e. you can't have two tests running a job called `example`, you need to keep in mind resource constraints on the cluster (it's not that big), etc. I'd recommend trying to make it _possible_ to run them in parallel but not doing it until we've got a handle on this a bit more first. Walk, then run. 😀 

> Tying into the previous question, what level of test coverage are we looking to accomplish? Do we want confidence in UI-specific features? Or do we want to test Nomad as a whole and fully mimic how our end users are configured and running Nomad?

The test environment is "realistic" in that it has ACLs and mTLS and a (small) diversity of clients. Because these tests can take a while, you want to get a lot of bang-for-buck: what are the UI workflows that are hardest to test with Ember acceptance tests?

---

All this discussion of configuration _does_ make me realize there's one more bit of environment we'll need in the test runner. When we're talking to the UI we can use just `NOMAD_ADDR` and `NOMAD_TOKEN` because we've got a [reverse proxy](https://github.com/hashicorp/nomad/blob/main/e2e/ui/run.sh#L18-L21) serving the UI to us. But when we're talking directly to the Nomad HTTP API, we'll also need to configure the HTTP API client for mTLS. 

That's going to be a little nastier to do and involve some messing with the infra to get this exposed to the test runner. If we merge this PR to `main` the tests will be broken, but it should be safe to merge to `trunk/playwright`. Why don't we do this?
* Let's merge this
* I'll take on the infrastructure bits we'll need and make a new PR to the `trunk/playwright` branch
* I'll mark you as a reviewer for that so that you can verify I'm giving the HTTP API client the right config values

Reply:
OK that sounds like a fair plan.

> Because these tests can take a while, you want to get a lot of bang-for-buck: what are the UI workflows that are hardest to test with Ember acceptance tests?

- All real-time data functionality. Updating charts, jobs, statuses.
- All features that use ACL + Namespaces.
- Accessing logs.
- The UI Exec Terminal feature.
- Error handling.
- One Time Tokens.
- and... technical debt that has us behind on upgrading to the latest version of `ember`. We are behind by 1 LTS and we're past the window where bug fixes for our version will continue coming out and soon we'll be passed the window for [security patches](https://emberjs.com/releases/lts/).